From 915cfc19aad135aba4a476d252dcbe23834905a6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Robert=20L=C3=B6hning?= <robert.loehning@qt.io>
Date: Fri, 17 Oct 2025 17:22:33 +0200
Subject: [PATCH 465/553] QPainterPath: Avoid division by zero in *AtPercent

Without the added checks, the code divided by the
paths' lengths which are zero.

Pick-to: 6.8
Change-Id: I78a2066af96136635f85f4937619ff398a28072c
Reviewed-by: Eirik Aavitsland <eirik.aavitsland@qt.io>
(cherry picked from commit 81f13dbadc428a044f9c75e9181928de08d2d8db)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/gui/painting/qpainterpath.cpp             | 11 ++++++++
 .../qpainterpath/tst_qpainterpath.cpp         | 26 +++++++++++++++++++
 2 files changed, 37 insertions(+)

diff --git a/src/gui/painting/qpainterpath.cpp b/src/gui/painting/qpainterpath.cpp
index 5763696266a..38c4a9ea87f 100644
--- a/src/gui/painting/qpainterpath.cpp
+++ b/src/gui/painting/qpainterpath.cpp
@@ -2937,6 +2937,8 @@ qreal QPainterPath::percentAtLength(qreal len) const
     if (len > totalLength)
         return 1;
 
+    Q_ASSERT(totalLength != 0);
+
     if (d->cacheEnabled) {
         const int ei = qMax(d->elementAtT(len / totalLength), 1); // Skip initial MoveTo
         qreal res = 0;
@@ -3117,6 +3119,7 @@ QPointF QPainterPath::pointAtPercent(qreal t) const
     qreal curLen = 0;
     qreal bezierLen = 0;
     QBezier b = d_ptr->bezierAtT(*this, t, &curLen, &bezierLen);
+    Q_ASSERT(bezierLen != 0);
     qreal realT = (totalLength * t - curLen) / bezierLen;
 
     return b.pointAt(qBound(qreal(0), realT, qreal(1)));
@@ -3141,10 +3144,14 @@ qreal QPainterPath::angleAtPercent(qreal t) const
         return 0;
     }
 
+    if (isEmpty())
+        return 0;
+
     qreal totalLength = length();
     qreal curLen = 0;
     qreal bezierLen = 0;
     QBezier bez = d_ptr->bezierAtT(*this, t, &curLen, &bezierLen);
+    Q_ASSERT(bezierLen != 0);
     qreal realT = (totalLength * t - curLen) / bezierLen;
 
     qreal m1 = slopeAt(realT, bez.x1, bez.x2, bez.x3, bez.x4);
@@ -3170,10 +3177,14 @@ qreal QPainterPath::slopeAtPercent(qreal t) const
         return 0;
     }
 
+    if (isEmpty())
+        return 0;
+
     qreal totalLength = length();
     qreal curLen = 0;
     qreal bezierLen = 0;
     QBezier bez = d_ptr->bezierAtT(*this, t, &curLen, &bezierLen);
+    Q_ASSERT(bezierLen != 0);
     qreal realT = (totalLength * t - curLen) / bezierLen;
 
     qreal m1 = slopeAt(realT, bez.x1, bez.x2, bez.x3, bez.x4);
diff --git a/tests/auto/gui/painting/qpainterpath/tst_qpainterpath.cpp b/tests/auto/gui/painting/qpainterpath/tst_qpainterpath.cpp
index bd9e00fc376..7b4193e3749 100644
--- a/tests/auto/gui/painting/qpainterpath/tst_qpainterpath.cpp
+++ b/tests/auto/gui/painting/qpainterpath/tst_qpainterpath.cpp
@@ -86,6 +86,9 @@ private slots:
     void intersectionPointOnEdge();
 
     void boundsAtStartPoint();
+
+    void percentAtEmptyPath_data();
+    void percentAtEmptyPath();
 };
 
 void tst_QPainterPath::cleanupTestCase()
@@ -1637,6 +1640,29 @@ void tst_QPainterPath::boundsAtStartPoint()
     QCOMPARE(constructedPath.controlPointRect(), defaultPath.controlPointRect());
 }
 
+void tst_QPainterPath::percentAtEmptyPath_data()
+{
+    const QPointF qpf(2., 2.);
+    QTest::addColumn<QPainterPath>("path");
+
+    QTest::newRow("defaultConstructed") << QPainterPath();
+    QTest::newRow("withStartPoint") << QPainterPath(qpf);
+    QPainterPath pp(qpf);
+    pp.lineTo(qpf);
+    QTest::newRow("equalPoints") << pp;
+}
+
+void tst_QPainterPath::percentAtEmptyPath()
+{
+    QFETCH(QPainterPath, path);
+
+    path.angleAtPercent(0.5);
+    path.pointAtPercent(0.5);
+    path.slopeAtPercent(0.5);
+
+    path.percentAtLength(0);
+}
+
 QTEST_APPLESS_MAIN(tst_QPainterPath)
 
 #include "tst_qpainterpath.moc"
-- 
2.51.2

